src/hotspot/os/posix/os_posix.cpp

```diff
@@ -548,6 +549,32 @@ void os::Posix::print_user_info(outputStream* st) {
   st->cr();
 }
 
+// Print all active locale categories, one line each
+void os::Posix::print_active_locale(outputStream* st) {
+  // Posix is quiet about how exactly LC_ALL is implemented.
+  // Just print it out too, in case LC_ALL is held separately
+  // from the individual categories.
+  #define LOCALE_CAT_DO(f) \
+    f(LC_ALL) \
+    f(LC_COLLATE) \
+    f(LC_CTYPE) \
+    f(LC_MESSAGES) \
+    f(LC_MONETARY) \
+    f(LC_NUMERIC) \
+    f(LC_TIME)
+  #define XX(cat) { cat, #cat },
+  const struct { int c; const char* name; } categories[] = {
+      LOCALE_CAT_DO(XX)
+      { -1, NULL }
+  };
+  #undef XX
+  #undef LOCALE_CAT_DO
+  for (int i = 0; categories[i].c != -1; i ++) {
+    const char* locale = setlocale(categories[i].c, NULL);
+    st->print_cr("%s=%s", categories[i].name,
+                 ((locale != NULL) ? locale : ""));
```

Comment:
I can't see where the query version of setLocale is allowed to return NULL, but if it did isn't that different to being set to "" ?

Reply:
"The return value is NULL if the request cannot be honored." linux manpage.
I can print "??" instead?

Reply:
The Linux manpage is even more vague than the Posix spec. :(

I'd prefer "<null>" or "<not-set>" or "<unknown>" to just "??".

Reply:
Sorry I never noticed that. I will add escape characters to make it clear for future readers :)